SSA/Projects/LLVMRiscV/simpproc.lean

```diff
@@ -0,0 +1,144 @@
+      throwError "Failed to unify {tys'} with {expected}"
+  let expr : Lean.Expr :=
+    let x : Q(⟦$ty⟧) := x
+    let xs : Q(HVector TyDenote.toType (List.map LLVMRiscV.Ty.llvm $tys)) := xs
+    q(HVector.cons (f := TyDenote.toType) (a := $ty) (as := $tys) $x (llvmArgsFromHybrid $xs))
+  let proof := mkAppN (mkConst ``llvmArgsFromHybrid_cons_eq.lemma []) #[ty, tys, x, xs]
+  return .visit { expr, proof? := some proof }
+
+@[simp_denote]
+def riscvArgsFromHybrid_nil_eq :
+  (riscvArgsFromHybrid HVector.nil) = HVector.nil := rfl
+@[simp_denote]
+def riscvArgsFromHybrid_cons_eq.lemma {ty  : RISCV64.RV64.Ty} {tys : List RISCV64.RV64.Ty}
+    (x : TyDenote.toType (LLVMRiscV.Ty.riscv ty))
+    (xs : HVector TyDenote.toType (tys.map LLVMRiscV.Ty.riscv)) :
+  (riscvArgsFromHybrid (tys := ty :: tys) (HVector.cons x xs)) =
+  HVector.cons (f := TyDenote.toType) (a := ty) (as := tys) x (riscvArgsFromHybrid xs)
+   := rfl
+
+open Lean Meta Elab in
+/-- Extracting out the raw LLVM type. -/
+def extractRiscvTy (x : Expr) : SimpM Expr := do
+  let_expr Ty.riscv xRealTy := (← reduce x)
+    | throwError m! "expected type of {x} to be `Ty.riscv _`, but got {x}"
+  return xRealTy
+
+open Lean Meta Elab in
+simproc [simp_denote] riscvArgsFromHybrid_cons_eq (riscvArgsFromHybrid _) := fun e => do
+  let_expr riscvArgsFromHybrid _ lhs := e
+    | throwError m!"unable to find riscvArgsFromHybrid in {e}"
+  let_expr HVector.cons _α _f as a x xs := lhs
+    | throwError m!"unable to find HVector.cons in {lhs}"
+  let xRealTy ← extractRiscvTy a
+  let some (_, xsRealTys) := Expr.listLit? (← reduce as)
+    | return .continue
+  let xsRealTys ←  xsRealTys.mapM extractRiscvTy
+  logInfo m!"found (llvmArgsFromHybrid (HVector.cons ({x} : {xRealTy}) ({xs} : {xsRealTys})"
+  let llvmTypeType := mkApp (mkConst ``Dialect.Ty []) (mkConst ``RISCV64.RV64 [])
+  let xsRealTys ← listExprToExprShallow (.some llvmTypeType) xsRealTys
+  logInfo m!"calling {``riscvArgsFromHybrid_cons_eq.lemma} with {xRealTy}, {xsRealTys}, {x}, {xs}"
+  logInfo m!"XXXX"
+  let proof := mkAppN (mkConst ``riscvArgsFromHybrid_cons_eq.lemma []) #[xRealTy, xsRealTys, x, xs]
+  logInfo m!"YYYY"
+  logInfo m!"built proof {proof}"
+  let proof ← reduce proof
+  logInfo m!"reduced proof to {proof}"
+  let eq ← reduce (← inferType proof)
+  logInfo m!"reduced type of proof (i.e. the equality) to {eq}"
+  let .some (_ty, _lhs, rhs) := eq.eq?
+    | throwError "unable to reduce application of riscvArgsFromHybrid_cons_eq.lemma to an
+      equality, only reduced to '{eq}'."
+  logInfo m!"final right-hand-side of equality is: {rhs}"
+  return .visit {
+    expr := rhs,
+    proof? := .some proof
+  }
+
+/- The following additional lemmas are needed to simplify the proof goals for the hybrid dialect.-/
+@[simp]
+theorem poisonOr_mk_some_eq_value (x : α) : { toOption := some x } = PoisonOr.value x := rfl
+
+@[simp_denote]
+theorem liftM_eq_some (α : Type u) : @liftM Id Option _  α = some := by rfl
```

Comment:
Do you still need this? I thought we got rid of the superfluous option that got introduced?